pkg/util/controllerfinder/controller_finder.go

```diff
@@ -365,22 +372,100 @@ func (r *ControllerFinder) getPodKruiseStatefulSet(ref ControllerReference, name
 	return &ScaleAndSelector{
 		Scale:           *(ss.Spec.Replicas),
 		ReserveOrdinals: ss.Spec.ReserveOrdinals,
-		Selector:        ss.Spec.Selector,
+		Selector:        metav1.FormatLabelSelector(ss.Spec.Selector),
 		ControllerReference: ControllerReference{
 			APIVersion: ss.APIVersion,
 			Kind:       ss.Kind,
 			Name:       ss.Name,
 			UID:        ss.UID,
 		},
-		Metadata:   ss.ObjectMeta,
-		TempLabels: ss.Spec.Template.Labels,
+		Metadata: ss.ObjectMeta,
+	}, nil
+}
+
+func (r *ControllerFinder) getScaleController(ref ControllerReference, namespace string) (*ScaleAndSelector, error) {
+	if isValidGroupVersionKind(ref.APIVersion, ref.Kind) {
+		return nil, nil
+	}
+	gv, err := schema.ParseGroupVersion(ref.APIVersion)
+	if err != nil {
+		return nil, err
+	}
+	gk := schema.GroupKind{
+		Group: gv.Group,
+		Kind:  ref.Kind,
+	}
+
+	mapping, err := r.mapper.RESTMapping(gk, gv.Version)
+	if err != nil {
+		return nil, err
+	}
+	gr := mapping.Resource.GroupResource()
+	scale, err := r.scaleNamespacer.Scales(namespace).Get(context.TODO(), gr, ref.Name, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// The IsNotFound error can mean either that the resource does not exist,
+			// or it exist but doesn't implement the scale subresource. We check which
+			// situation we are facing so we can give an appropriate error message.
+			isScale, err := r.implementsScale(gv, ref.Kind)
+			if err != nil {
+				return nil, err
+			}
+			if !isScale {
+				return nil, fmt.Errorf("%s does not implement the scale subresource", gr.String())
+			}
+			return nil, nil
+		}
+		return nil, err
+	}
+	if ref.UID != "" && scale.UID != ref.UID {
+		return nil, nil
+	}
+	return &ScaleAndSelector{
+		Scale: scale.Spec.Replicas,
+		ControllerReference: ControllerReference{
+			APIVersion: ref.APIVersion,
+			Kind:       ref.Kind,
+			Name:       ref.Name,
+			UID:        scale.UID,
+		},
+		Metadata: scale.ObjectMeta,
+		Selector: scale.Status.Selector,
 	}, nil
 }
 
+func (r *ControllerFinder) implementsScale(gv schema.GroupVersion, kind string) (bool, error) {
+	resourceList, err := r.discoveryClient.ServerResourcesForGroupVersion(gv.String())
+	if err != nil {
+		return false, err
+	}
+	for _, resource := range resourceList.APIResources {
```

Comment:
The logic here is incorrect, please fix it according to this https://github.com/kubernetes/kubernetes/pull/109957